sydent/http/servlets/replication.py

```diff
@@ -40,67 +69,139 @@ def render_POST(self, request):
+                    request.finish()
+                    return
 
                 assocObj = threePidAssocFromDict(sgAssoc)
 
                 if assocObj.mxid is not None:
+                    # Add the association components and the original signed
+                    # object (as assocs must be signed when requested by clients)
                     globalAssocsStore.addAssociation(assocObj, json.dumps(sgAssoc), peer.servername, originId, commit=False)
                 else:
                     logger.info("Incoming deletion: removing associations for %s / %s", assocObj.medium, assocObj.address)
                     globalAssocsStore.removeAssociation(assocObj.medium, assocObj.address)
-                logger.info("Stored association origin ID %s from %s", originId, peer.servername)
+
+                logger.info("Stored association with origin ID %s from %s", originId, peer.servername)
 
                 # if this is an association that matches one of our invite_tokens then we should call the onBind callback
                 # at this point, in order to tell the inviting HS that someone out there has just bound the 3PID.
                 self.sydent.threepidBinder.notifyPendingInvites(assocObj)
 
-            except:
-                failedIds.append(originId)
-                logger.warn("Failed to verify signed association from %s with origin ID %s",
-                            peer.servername, originId)
-                twisted.python.log.err()
-
-        if len(failedIds) > 0:
-            self.sydent.db.rollback()
-            request.setResponseCode(400)
-            return {'errcode': 'M_VERIFICATION_FAILED', 'error': 'Verification failed for one or more associations',
-                    'failed_ids':failedIds}
-        else:
-            self.sydent.db.commit()
-            return {'success':True}
+        # Check for any invite tokens
+        if 'invite_tokens' in inJson or 'ephemeral_public_keys' in inJson:
+            tokensStore = JoinTokenStore(self.sydent)
+
+            if 'invite_tokens' in inJson and len(inJson['invite_tokens']) > 0:
+                if len(inJson['invite_tokens']) > MAX_INVITE_TOKENS_LIMIT:
+                    self.sydent.db.rollback()
+                    request.setResponseCode(400)
+                    request.write(json.dumps({'errcode': 'M_BAD_JSON', 'error': '"invite_tokens" has more than %d keys' % MAX_INVITE_TOKENS_LIMIT}))
+                    request.finish()
+                    return
+
+                last_processed_id = tokensStore.getLastTokenIdFromServer(peer.servername)
+                for originId, inviteToken in inJson["invite_tokens"].items():
+                    # Make sure we haven't processed this token already
+                    # If so, back out of all incoming tokens and return an error
+                    if int(originId) <= int(last_processed_id):
+                        self.sydent.db.rollback()
+                        request.setResponseCode(200)
```

Comment:
this doesn't look like an error. 

Are you sure it's correct that we abort in this case rather than just ignoring it?

Reply:
(and if we're going to ignore it, could we move that logic into the tokensStore?)

Reply:
We return a 200 if we're already processed some IDs, such that the sending server doesn't continuously try to send the same batch over and over again.

Reply:
And yeah, we can move the logic.

Reply:
Returning a 200 sounds like a sensible thing to do. ignoring everything else in the request doesn't sound like  a sensible thing to do. Having a comment that says "return an error" just before we return a 200 doesn't sound like a sensible thing to do.

I'm also not sure if the `message` that we return will be expected by anything.